nni/algorithms/hpo/tpe_tuner.py

```diff
@@ -0,0 +1,374 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+"""
+Tree-structured Parzen Estimator (TPE) tuner for hyper-parameter optimization.
+
+Paper: https://proceedings.neurips.cc/paper/2011/file/86e8f7ab32cfd12577bc2619bc635690-Paper.pdf
+Official code: https://github.com/hyperopt/hyperopt/blob/master/hyperopt/tpe.py
+
+This is a slightly modified re-implementation of the algorithm.
+"""
+
+__all__ = ['TpeTuner', 'TpeArguments', 'suggest', 'suggest_parameter']
+
+from collections import defaultdict
+import logging
+import math
+from typing import NamedTuple, Optional, Union
+
+import numpy as np
+from scipy.special import erf  # pylint: disable=no-name-in-module
+
+from nni.tuner import Tuner
+from nni.common.hpo_utils import OptimizeMode, format_search_space, deformat_parameters
+from . import random_tuner
+
+_logger = logging.getLogger('nni.tuner.tpe')
+
+## Public API part ##
+
+class TpeArguments(NamedTuple):
+    """
+    These are the hyper-parameters of TPE algorithm itself.
+    To avoid confusing with trials' hyper-parameters, they are called "arguments" in this code.
+
+    Parameters
+    ==========
+    constant_liar_type: 'best' | 'worst' | 'mean' | None (default: 'best')
+        TPE algorithm itself does not support parallel tuning.
+        This parameter specifies how to optimize for trial_concurrency > 1.
+
+        None (or "null" in YAML) means do not optimize. This is the default behavior in legacy version.
+
+        How each liar works is explained in paper's section 6.1.
+        In general "best" suit for small trial number and "worst" suit for large trial number.
+
+    n_startup_jobs: int (default: 20)
+        The first N hyper-parameters are generated fully randomly for warming up.
+        If the search space is large, you can increase this value.
+        Or if max_trial_number is small, you man want to decrease it.
+
+    n_ei_candidates: int (default: 24)
+        For each iteration TPE samples EI for N sets of parameters and choose the best one. (loosely speaking)
+
+    linear_forgetting: int (default: 25)
+        TPE will lower the weights of old trials.
+        This controls how many iterations it takes for a trial to start decay.
+
+    prior_weight: float (default: 1.0)
+        TPE uses history trials plus the search space itself (called "prior") to choose parameters.
+        With prior weight 1.0, the search space is treated as one good trial.
+        For example, "normal(0, 1)" effectly equals to a trial with x = 0 which has yielded good result.
```

Comment:
do not understand... could you explain more?

Reply:
It's really hard to express...
For example, you have a search space `x = normal(0, 1)`.
And you have run 5 trials. 2 of them have "good" result (see `gamma`): `x=-0.5`, `x=0.1`;
the other 3 have "bad" result: `x=0.1`, `x=-0.4`, `x=1.1`.

Then, TPE starts to make suggestion. It builds internal state base on "good_x" and "bad_x".
With `prior_weight = 0`, the state is something like: `good_x = [-0.5, 0.1] ; bad_x = [-0.4, 0.1, 1.1]`.
With `prior_weight = 1`, the state is like: `good_x = [-0.5, 0, 0.1] ; bad_x = [-0.4, 0.1, 1.1]`.
With `prior_weight = 2`, the state is like: `good_x = [-0.5, 0, 0, 0.1] ; bad_x = [-0.4, 0.1, 1.1]`.

Above states are simplified. In real implementation each trial has a weight (see `linear_forgetting`), and that's why `prior_weight` is a float rather than int.

Can you help me rewrite the explanation in short words?

Reply:
why adding 0 to good_x rather than other values, when prior_weight >0?

Reply:
Because in the search space mu=0.

Reply:
TPE treats user provided search space as prior. When generating new trials, it also incorporates the prior in trial history by transforming the search space to one trial configuration (i.e., each hyper-parameter of this configuration chooses the mean of its candidate range). Here, prior_weight determines the weight of this trial configuration in the history trial configurations. With prior weight 1.0, the search space is treated as one good trial. For example, "normal(0, 1)" effectly equals to a trial with x = 0 which has yielded good result.